### PR TITLE
Add checks config fields and get_checks API

### PR DESCRIPTION
**Pull request by benhoyt**
Add support for the new backoff and checks-related fields in the plan configuration types, as well as support for the new `/v1/checks` API (`Client.get_checks` and `Container.get_checks` and `Container.get_check`).

See also the brief docs I'm adding in https://github.com/benhoyt/juju-docs/pull/3 (in addition to the API reference docs in this PR).

### CODE REVIEW

**Review by hpidcock — APPROVED** [review]
LGTM

**Review by jnsgruk — APPROVED** [review]
Nice work, thanks!

**Review by sed-i — COMMENTED** [review]
A couple of type hints for consistency

File: ops/pebble.py
Comment (benhoyt) [reply ops/pebble.py]: Done.

File: ops/pebble.py
Comment (benhoyt) [reply ops/pebble.py]: Done.

**Review by niemeyer — CHANGES_REQUESTED** [review]
On a quick skim, this is looking nice, thanks!

Let's please hold until the server is merged as there will be minor differences.

**Comment by rwcarlsen** [comment]
I assume we're waiting on this: https://github.com/canonical/pebble/pull/86?

**Comment by benhoyt** [comment]
@rwcarlsen Yep, that's right. I'll note that in the PR description.

**Comment by benhoyt** [comment]
@rwcarlsen Yep, here they are: https://github.com/benhoyt/juju-docs/pull/1 ... the plan is to add them (and Dora will review/edit) once https://github.com/canonical/pebble/pull/86 and then this PR is merged.

**Comment by jnsgruk** [comment]
@benhoyt Looking at activity today, looks like we're pretty darn close to getting this merged? :)

File: .github/workflows/observability-charm-tests.yaml
Comment (benhoyt) [inline .github/workflows/observability-charm-tests.yaml:26]: We can remove this later when https://github.com/canonical/grafana-k8s-operator/pull/69 is merged, but it makes our CI fail so is needed to merge.

Comment (benhoyt) [reply .github/workflows/observability-charm-tests.yaml]: This is merged now, reverting this hack.

**Comment by rwcarlsen** [comment]
@benhoyt - have things settled enough here for another review?

**Comment by benhoyt** [comment]
@rwcarlsen Yeah, definitely. The Pebble changes are now merged - go ahead and review. Also I think we're waiting on @niemeyer's review here, as this is an API change (though that probably won't be before next week, given the sprint).

File: test/test_pebble.py
Comment (rwcarlsen) [inline test/test_pebble.py:3015]: Maybe this is a dumb question - but `sleep x` should exit nonzero right - so why does your assert below say the status is `UP`?

**Review by rwcarlsen — APPROVED** [review]
Looks good.  Looks like we'll need to make an ops framework guide doc for health checks.  Unless you have something in the works already - let's create an issue to track that.

File: ops/pebble.py
Comment (PietroPasotti) [inline ops/pebble.py:733]: Why the choice to make raw an untyped dict instead of unpacking it?
I would rather have `Check('name', override:str='', level:str='', period:str='', ...)`, makes for a way easier to read initializer, and easier usage of the class.
If you really dislike doing `check=Check('name', **my_raw_dict)`, then we should expose an alternative classmethod
`Check.from_dict(name:str, dict_:dict)`.

If we were on more modern python I could also go for a dataclass or a TypedDict, but until then, this feels like better code.

Comment (PietroPasotti) [reply ops/pebble.py:733]: Actually the whole Check class could very well be a dataclass, and save some __eq__, __repr__ and to_dict boilerplate

File: ops/pebble.py
Comment (PietroPasotti) [inline ops/pebble.py:894]: Same here. Or is this 'legacy code style'?

File: test/test_pebble.py
Comment (benhoyt) [reply test/test_pebble.py:3012]: It's only marked as `DOWN` after two checks have taken place (and the number of failures has hit the threshold). Which happens after 2 times the 50ms period. So initially the check will be `UP`, and only down after two tries, which is tested in the loop below. I've added comments here to clarify -- thanks.

File: ops/pebble.py
Comment (benhoyt) [reply ops/pebble.py:737]: Yeah, this should arguably be done with a `from_dict`. However, I've made this consistent with the other layer-configuration classes, which all do it the same way. If doing it again maybe we could have considered dataclasses or some such (then again, that's 3.7+ and `ops` still needs to support 3.5 for now). So I'm going to leave as is to be consistent.

File: ops/pebble.py
Comment (benhoyt) [reply ops/pebble.py:914]: See above.

**Comment by jnsgruk** [comment]
@benhoyt would appreciate if you could add the docs you drafted at some point! Thanks for all your work on this! :D

**Comment by benhoyt** [comment]
Will do on Monday, thanks @jnsgruk.

**Comment by benhoyt** [comment]
Added brief docs [here](https://juju.is/docs/sdk/pebble#heading--fetch-check-status). I'll ask Dora to review.

**Pull request merged**
